src/Service/Authenticator/ResourceAccess.php

```diff
@@ -0,0 +1,87 @@
+<?php
+/**
+ * Copyright Shopgate Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * @author    Shopgate Inc, 804 Congress Ave, Austin, Texas 78701 <interfaces@shopgate.com>
+ * @copyright Shopgate Inc
+ * @license   http://www.apache.org/licenses/LICENSE-2.0 Apache License, Version 2.0
+ */
+
+namespace Shopgate\CloudIntegrationSdk\Service\Authenticator;
+
+use Shopgate\CloudIntegrationSdk\Repository;
+use Shopgate\CloudIntegrationSdk\ValueObject\Request;
+use Shopgate\CloudIntegrationSdk\ValueObject\TokenId;
+use Shopgate\CloudIntegrationSdk\ValueObject\TokenType;
+use Shopgate\CloudIntegrationSdk\ValueObject\UserId;
+
+class ResourceAccess implements AuthenticatorInterface
+{
+    /** @var Repository\AbstractToken */
+    private $repository;
+
+    /** @var UserId  */
+    private $resourceOwnerId;
+
+    /**
+     * ResourceAdminAccess constructor.
+     *
+     * @param Repository\AbstractToken $accessTokenRepository
+     * @param UserId |null             $resourceOwnerId
+     */
+    public function __construct(Repository\AbstractToken $accessTokenRepository, UserId $resourceOwnerId = null) {
+        $this->repository = $accessTokenRepository;
+        $this->resourceOwnerId = $resourceOwnerId;
+    }
+
+    /**
+     * @param Request\Request $request
+     *
+     * @throws Exception\Unauthorized
+     * @throws Request\Exception\BadRequest
+     */
+    public function authenticate(Request\Request $request)
+    {
+        // read access_token parameter from header
+        $authHeader = $request->getHeader('Authorization');
+        if (empty($authHeader)) {
```

Comment:
In this case, it would not be fully correct, because the Request class could be extended and behave differently. So for some systems this might be not "null" and the authenticator to working properly in this case.